File: SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Analytics/ChartDataAggregator.swift
import Foundation

// MARK: - Time Range

/// Time ranges for analytics chart filtering.
public enum TimeRange: String, CaseIterable, Sendable {
    case lastMonth
    case lastThreeMonths
    case lastSixMonths
    case lastYear
    case allTime

    /// The start date for this time range, computed relative to the current date.
    public var startDate: Date {
        let calendar = Calendar.current
        let now = Date()
        switch self {
        case .lastMonth:
            return calendar.date(byAdding: .month, value: -1, to: now) ?? now
        case .lastThreeMonths:
            return calendar.date(byAdding: .month, value: -3, to: now) ?? now
        case .lastSixMonths:
            return calendar.date(byAdding: .month, value: -6, to: now) ?? now
        case .lastYear:
            return calendar.date(byAdding: .year, value: -1, to: now) ?? now
        case .allTime:
            return Date.distantPast
        }
    }

    /// Start date computed relative to a specific reference date (useful for testing).
    public func startDate(relativeTo reference: Date) -> Date {
        let calendar = Calendar.current
        switch self {
        case .lastMonth:
            return calendar.date(byAdding: .month, value: -1, to: reference) ?? reference
        case .lastThreeMonths:
            return calendar.date(byAdding: .month, value: -3, to: reference) ?? reference
        case .lastSixMonths:
            return calendar.date(byAdding: .month, value: -6, to: reference) ?? reference
        case .lastYear:
            return calendar.date(byAdding: .year, value: -1, to: reference) ?? reference
        case .allTime:
            return Date.distantPast
        }
    }
}

// MARK: - Data Points

/// A single data point for strength progression charts.
public struct StrengthDataPoint: Equatable, Sendable, Identifiable {
    public let id = UUID()
    public let date: Date
    public let exerciseName: String
    public let weight: Double
    public let unit: WeightUnit

    public init(date: Date, exerciseName: String, weight: Double, unit: WeightUnit) {
        self.date = date
        self.exerciseName = exerciseName
        self.weight = weight
        self.unit = unit
    }
}

/// A single data point for training volume charts, aggregated by week.
public struct VolumeDataPoint: Equatable, Sendable, Identifiable {
    public let id = UUID()
    public let weekStartDate: Date
    public let totalVolume: Double
    public let workoutCount: Int

    public init(weekStartDate: Date, totalVolume: Double, workoutCount: Int) {
        self.weekStartDate = weekStartDate
        self.totalVolume = totalVolume
        self.workoutCount = workoutCount
    }
}

/// A single data point for workout frequency charts, aggregated by week.
public struct FrequencyDataPoint: Equatable, Sendable, Identifiable {
    public let id = UUID()
    public let weekStartDate: Date
    public let workoutCount: Int
    public let label: String

    public init(weekStartDate: Date, workoutCount: Int, label: String) {
        self.weekStartDate = weekStartDate
        self.workoutCount = workoutCount
        self.label = label
    }
}

/// A single data point for cycle-correlated performance charts.
public struct CyclePerformancePoint: Equatable, Sendable, Identifiable {
    public var id: String { phase.rawValue }
    public let phase: CyclePhase
    public let averageVolume: Double
    public let workoutCount: Int
    public let confidence: Double

    public init(phase: CyclePhase, averageVolume: Double, workoutCount: Int, confidence: Double) {
        self.phase = phase
        self.averageVolume = averageVolume
        self.workoutCount = workoutCount
        self.confidence = confidence
    }
}

// MARK: - Aggregator

/// Pure domain service that transforms raw records into chart-ready data points.
///
/// All methods are static and side-effect free. The ViewModel layer calls these
/// with data fetched from the persistence store.
public enum ChartDataAggregator {

    // MARK: - Cached Formatters

    private static let weekLabelFormatter: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "MMM d"
        return f
    }()

    // MARK: Strength Progression

    /// Groups one-rep-max records by exercise, filters by time range, and sorts by date.
    ///
    /// - Parameters:
    ///   - records: Raw one-rep-max records from the persistence store.
    ///   - timeRange: The time window to filter by.
    ///   - referenceDate: Optional reference date for time range computation (defaults to now).
    /// - Returns: Sorted array of strength data points within the time range.
    public static func strengthProgression(
        from records: [OneRepMaxRecord],
        timeRange: TimeRange,
        referenceDate: Date = Date()
    ) -> [StrengthDataPoint] {
        let cutoff = timeRange.startDate(relativeTo: referenceDate)

        return records
            .filter { $0.date >= cutoff }
            .map { StrengthDataPoint(
                date: $0.date,
                exerciseName: $0.exerciseName,
                weight: $0.weight,
                unit: $0.unit
            ) }
            .sorted { $0.date < $1.date }
    }

    // MARK: Training Volume

    /// Aggregates completed workout volumes by ISO week.
    ///
    /// Only workouts with a non-nil `completedAt` are included.
    ///
    /// - Parameters:
    ///   - workouts: Raw workouts from the persistence store.
    ///   - timeRange: The time window to filter by.
    ///   - referenceDate: Optional reference date for time range computation (defaults to now).
    /// - Returns: Sorted array of weekly volume data points.
    public static func trainingVolume(
        from workouts: [Workout],
        timeRange: TimeRange,
        referenceDate: Date = Date()
    ) -> [VolumeDataPoint] {
        let cutoff = timeRange.startDate(relativeTo: referenceDate)
        let calendar = Calendar.current

        let completed = workouts.filter { workout in
            guard let completedAt = workout.completedAt else { return false }
            return completedAt >= cutoff
        }

        // Group by ISO week start date
        var weekMap: [Date: (volume: Double, count: Int)] = [:]

        for workout in completed {
            let weekStart = calendar.date(from: calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: workout.completedAt!))!
            if var existing = weekMap[weekStart] {
                existing.volume += workout.totalVolume
                existing.count += 1
                weekMap[weekStart] = existing
            } else {
                weekMap[weekStart] = (volume: workout.totalVolume, count: 1)
            }
        }

        return weekMap.map { key, value in
            VolumeDataPoint(weekStartDate: key, totalVolume: value.volume, workoutCount: value.count)
        }.sorted { $0.weekStartDate < $1.weekStartDate }
    }

    // MARK: Workout Frequency

    /// Counts completed workouts per ISO week.
    ///
    /// Only workouts with a non-nil `completedAt` are included.
    ///
    /// - Parameters:
    ///   - workouts: Raw workouts from the persistence store.
    ///   - timeRange: The time window to filter by.
    ///   - referenceDate: Optional reference date for time range computation (defaults to now).
    /// - Returns: Sorted array of weekly frequency data points with human-readable labels.
    public static func workoutFrequency(
        from workouts: [Workout],
        timeRange: TimeRange,
        referenceDate: Date = Date()
    ) -> [FrequencyDataPoint] {
        let cutoff = timeRange.startDate(relativeTo: referenceDate)
        let calendar = Calendar.current

        let completed = workouts.filter { workout in
            guard let completedAt = workout.completedAt else { return false }
            return completedAt >= cutoff
        }

        var weekMap: [Date: Int] = [:]

        for workout in completed {
            let weekStart = calendar.date(from: calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: workout.completedAt!))!
            weekMap[weekStart, default: 0] += 1
        }

        return weekMap.map { key, count in
            let label = "Week of \(weekLabelFormatter.string(from: key))"
            return FrequencyDataPoint(weekStartDate: key, workoutCount: count, label: label)
        }.sorted { $0.weekStartDate < $1.weekStartDate }
    }

    // MARK: Cycle Correlation

    /// Maps workouts to their cycle phases via date overlap and averages volume per phase.
    ///
    /// Each workout is matched to the first `CyclePhaseInfo` whose date range contains
    /// the workout's `completedAt` date. The resulting points include the average
    /// confidence of matched phases.
    ///
    /// - Parameters:
    ///   - workouts: Raw workouts from the persistence store.
    ///   - phases: Cycle phase info records with date ranges.
    ///   - timeRange: The time window to filter by.
    ///   - referenceDate: Optional reference date for time range computation (defaults to now).
    /// - Returns: Array of performance points, one per represented phase.
    public static func cycleCorrelation(
        from workouts: [Workout],
        phases: [CyclePhaseInfo],
        timeRange: TimeRange,
        referenceDate: Date = Date()
    ) -> [CyclePerformancePoint] {
        let cutoff = timeRange.startDate(relativeTo: referenceDate)
        let calendar = Calendar.current

        let completed = workouts.filter { workout in
            guard let completedAt = workout.completedAt else { return false }
            return completedAt >= cutoff
        }

        // Map workouts to phases
        var phaseData: [CyclePhase: (volumes: [Double], confidences: [Double])] = [:]

        for workout in completed {
            let workoutDate = workout.completedAt!
            let dayStart = calendar.startOfDay(for: workoutDate)

            // Find the first phase whose date range contains this workout
            guard let matchingPhase = phases.first(where: { phase in
                let phaseStart = calendar.startOfDay(for: phase.startDate)
                let phaseEnd = calendar.startOfDay(for: phase.endDate)
                return dayStart >= phaseStart && dayStart <= phaseEnd
            }) else {
                continue
            }

            if var existing = phaseData[matchingPhase.phase] {
                existing.volumes.append(workout.totalVolume)
                existing.confidences.append(matchingPhase.confidence)
                phaseData[matchingPhase.phase] = existing
            } else {
                phaseData[matchingPhase.phase] = (
                    volumes: [workout.totalVolume],
                    confidences: [matchingPhase.confidence]
                )
            }
        }

        return phaseData.map { phase, data in
            let avgVolume = data.volumes.reduce(0, +) / Double(data.volumes.count)
            let avgConfidence = data.confidences.reduce(0, +) / Double(data.confidences.count)
            return CyclePerformancePoint(
                phase: phase,
                averageVolume: avgVolume,
                workoutCount: data.volumes.count,
                confidence: avgConfidence
            )
        }.sorted { $0.phase.rawValue < $1.phase.rawValue }
    }

    // MARK: Exercise Picker

    /// Returns unique exercise names from one-rep-max records, sorted alphabetically.
    ///
    /// Useful for populating the exercise picker in the strength progression chart.
    ///
    /// - Parameter records: Raw one-rep-max records.
    /// - Returns: Sorted array of unique exercise names.
    public static func exercises(from records: [OneRepMaxRecord]) -> [String] {
        Array(Set(records.map(\.exerciseName))).sorted()
    }
}

File: android/core/src/main/java/com/sundeefundee/core/domain/analytics/ChartDataAggregator.kt
package com.sundeefundee.core.domain.analytics

import com.sundeefundee.core.domain.benchmark.BenchmarkResult
import com.sundeefundee.core.model.CyclePhase
import com.sundeefundee.core.model.CyclePhaseInfo
import com.sundeefundee.core.model.OneRepMaxRecord
import com.sundeefundee.core.model.WeightUnit
import com.sundeefundee.core.model.Workout
import kotlinx.datetime.Clock
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDate
import kotlinx.datetime.TimeZone
import kotlinx.datetime.minus
import kotlinx.datetime.toLocalDateTime
import java.util.UUID

// MARK: - Time Range

/**
 * Time ranges for analytics chart filtering.
 */
enum class TimeRange {
    LAST_MONTH,
    LAST_THREE_MONTHS,
    LAST_SIX_MONTHS,
    LAST_YEAR,
    ALL_TIME;

    /** The start instant for this time range, computed relative to now. */
    val startInstant: Instant
        get() = startInstantRelativeTo(Clock.System.now())

    /** Start instant computed relative to a specific reference instant (useful for testing). */
    fun startInstantRelativeTo(reference: Instant): Instant {
        val tz = TimeZone.currentSystemDefault()
        return when (this) {
            LAST_MONTH -> reference.minus(1, DateTimeUnit.MONTH, tz)
            LAST_THREE_MONTHS -> reference.minus(3, DateTimeUnit.MONTH, tz)
            LAST_SIX_MONTHS -> reference.minus(6, DateTimeUnit.MONTH, tz)
            LAST_YEAR -> reference.minus(1, DateTimeUnit.YEAR, tz)
            ALL_TIME -> Instant.DISTANT_PAST
        }
    }
}

// MARK: - Data Points

/**
 * A single data point for strength progression charts.
 */
data class StrengthDataPoint(
    val id: String = UUID.randomUUID().toString(),
    val date: Instant,
    val exerciseName: String,
    val weight: Double,
    val unit: WeightUnit
)

/**
 * A single data point for training volume charts, aggregated by week.
 */
data class VolumeDataPoint(
    val id: String = UUID.randomUUID().toString(),
    val weekStartDate: LocalDate,
    val totalVolume: Double,
    val workoutCount: Int
)

/**
 * A single data point for workout frequency charts, aggregated by week.
 */
data class FrequencyDataPoint(
    val id: String = UUID.randomUUID().toString(),
    val weekStartDate: LocalDate,
    val workoutCount: Int,
    val label: String
)

/**
 * A single data point for cycle-correlated performance charts.
 */
data class CyclePerformancePoint(
    val phase: CyclePhase,
    val averageVolume: Double,
    val workoutCount: Int,
    val confidence: Double
) {
    val id: String get() = phase.name
}

// MARK: - Aggregator

/**
 * Pure domain service that transforms raw records into chart-ready data points.
 *
 * All functions are top-level and side-effect free. The ViewModel layer calls these
 * with data fetched from the persistence store.
 */
object ChartDataAggregator {

    // MARK: - Strength Progression

    /**
     * Groups one-rep-max records by exercise, filters by time range, and sorts by date.
     *
     * @param records Raw one-rep-max records from the persistence store.
     * @param timeRange The time window to filter by.
     * @param referenceDate Optional reference instant for time range computation (defaults to now).
     * @return Sorted list of strength data points within the time range.
     */
    fun strengthProgression(
        records: List<OneRepMaxRecord>,
        timeRange: TimeRange,
        referenceDate: Instant = Clock.System.now()
    ): List<StrengthDataPoint> {
        val cutoff = timeRange.startInstantRelativeTo(referenceDate)

        return records
            .filter {
                val instant = Instant.parse(it.date)
                instant >= cutoff
            }
            .map {
                StrengthDataPoint(
                    date = Instant.parse(it.date),
                    exerciseName = it.exerciseName,
                    weight = it.weight,
                    unit = it.unit
                )
            }
            .sortedBy { it.date }
    }

    // MARK: - Training Volume

    /**
     * Aggregates completed workout volumes by ISO week.
     *
     * Only workouts with a non-null [Workout.completedAt] are included.
     *
     * @param workouts Raw workouts from the persistence store.
     * @param timeRange The time window to filter by.
     * @param referenceDate Optional reference instant for time range computation (defaults to now).
     * @return Sorted list of weekly volume data points.
     */
    fun trainingVolume(
        workouts: List<Workout>,
        timeRange: TimeRange,
        referenceDate: Instant = Clock.System.now()
    ): List<VolumeDataPoint> {
        val cutoff = timeRange.startInstantRelativeTo(referenceDate)
        val tz = TimeZone.currentSystemDefault()

        val completed = workouts.filter { workout ->
            val completedAt = workout.completedAt ?: return@filter false
            Instant.parse(completedAt) >= cutoff
        }

        // Group by ISO week start date
        val weekMap = mutableMapOf<LocalDate, Pair<Double, Int>>()

        for (workout in completed) {
            val completedInstant = Instant.parse(workout.completedAt!!)
            val weekStart = isoWeekStart(completedInstant, tz)

            val existing = weekMap[weekStart]
            if (existing != null) {
                weekMap[weekStart] = (existing.first + workout.totalVolume) to (existing.second + 1)
            } else {
                weekMap[weekStart] = workout.totalVolume to 1
            }
        }

        return weekMap.map { (weekStart, data) ->
            VolumeDataPoint(
                weekStartDate = weekStart,
                totalVolume = data.first,
                workoutCount = data.second
            )
        }.sortedBy { it.weekStartDate }
    }

    // MARK: - Workout Frequency

    /**
     * Counts completed workouts per ISO week.
     *
     * Only workouts with a non-null [Workout.completedAt] are included.
     *
     * @param workouts Raw workouts from the persistence store.
     * @param timeRange The time window to filter by.
     * @param referenceDate Optional reference instant for time range computation (defaults to now).
     * @return Sorted list of weekly frequency data points with human-readable labels.
     */
    fun workoutFrequency(
        workouts: List<Workout>,
        timeRange: TimeRange,
        referenceDate: Instant = Clock.System.now()
    ): List<FrequencyDataPoint> {
        val cutoff = timeRange.startInstantRelativeTo(referenceDate)
        val tz = TimeZone.currentSystemDefault()

        val completed = workouts.filter { workout ->
            val completedAt = workout.completedAt ?: return@filter false
            Instant.parse(completedAt) >= cutoff
        }

        val weekMap = mutableMapOf<LocalDate, Int>()

        for (workout in completed) {
            val completedInstant = Instant.parse(workout.completedAt!!)
            val weekStart = isoWeekStart(completedInstant, tz)
            weekMap[weekStart] = (weekMap[weekStart] ?: 0) + 1
        }

        return weekMap.map { (weekStart, count) ->
            val label = formatWeekLabel(weekStart)
            FrequencyDataPoint(
                weekStartDate = weekStart,
                workoutCount = count,
                label = label
            )
        }.sortedBy { it.weekStartDate }
    }

    // MARK: - Cycle Correlation

    /**
     * Maps workouts to their cycle phases via date overlap and averages volume per phase.
     *
     * Each workout is matched to the first [CyclePhaseInfo] whose date range contains
     * the workout's [Workout.completedAt] date. The resulting points include the average
     * confidence of matched phases.
     *
     * @param workouts Raw workouts from the persistence store.
     * @param phases Cycle phase info records with date ranges.
     * @param timeRange The time window to filter by.
     * @param referenceDate Optional reference instant for time range computation (defaults to now).
     * @return List of performance points, one per represented phase.
     */
    fun cycleCorrelation(
        workouts: List<Workout>,
        phases: List<CyclePhaseInfo>,
        timeRange: TimeRange,
        referenceDate: Instant = Clock.System.now()
    ): List<CyclePerformancePoint> {
        val cutoff = timeRange.startInstantRelativeTo(referenceDate)
        val tz = TimeZone.currentSystemDefault()

        val completed = workouts.filter { workout ->
            val completedAt = workout.completedAt ?: return@filter false
            Instant.parse(completedAt) >= cutoff
        }

        // Map workouts to phases
        val phaseData = mutableMapOf<CyclePhase, Pair<MutableList<Double>, MutableList<Double>>>()

        for (workout in completed) {
            val workoutInstant = Instant.parse(workout.completedAt!!)
            val workoutLocalDate = workoutInstant.toLocalDateTime(tz).date

            // Find the first phase whose date range contains this workout
            val matchingPhase = phases.firstOrNull { phase ->
                val phaseStart = LocalDate.parse(phase.startDate)
                val phaseEnd = LocalDate.parse(phase.endDate)
                workoutLocalDate >= phaseStart && workoutLocalDate <= phaseEnd
            } ?: continue

            val existing = phaseData[matchingPhase.phase]
            if (existing != null) {
                existing.first.add(workout.totalVolume)
                existing.second.add(matchingPhase.confidence)
            } else {
                phaseData[matchingPhase.phase] = Pair(
                    mutableListOf(workout.totalVolume),
                    mutableListOf(matchingPhase.confidence)
                )
            }
        }

        return phaseData.map { (phase, data) ->
            val avgVolume = data.first.sum() / data.first.size
            val avgConfidence = data.second.sum() / data.second.size
            CyclePerformancePoint(
                phase = phase,
                averageVolume = avgVolume,
                workoutCount = data.first.size,
                confidence = avgConfidence
            )
        }.sortedBy { it.phase.name }
    }

    // MARK: - Exercise Picker

    /**
     * Returns unique exercise names from one-rep-max records, sorted alphabetically.
     *
     * Useful for populating the exercise picker in the strength progression chart.
     *
     * @param records Raw one-rep-max records.
     * @return Sorted list of unique exercise names.
     */
    fun exercises(records: List<OneRepMaxRecord>): List<String> {
        return records.map { it.exerciseName }.toSortedSet().toList()
    }

    // MARK: - Private Helpers

    /** Get the ISO week start date (Monday) for an instant. */
    private fun isoWeekStart(instant: Instant, tz: TimeZone): LocalDate {
        val localDate = instant.toLocalDateTime(tz).date
        val dow = localDate.dayOfWeek
        // Iso8601: Monday = 1, Sunday = 7
        val daysSinceMonday = dow.ordinal // Monday=0 ... Sunday=6
        return localDate.minus(daysSinceMonday, DateTimeUnit.DAY)
    }

    /** Format a week start date as "MMM d" for chart labels. */
    private fun formatWeekLabel(weekStart: LocalDate): String {
        val months = listOf("Jan", "Feb", "Mar", "Apr", "May", "Jun",
            "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
        val monthName = months[weekStart.monthNumber - 1]
        return "Week of $monthName ${weekStart.dayOfMonth}"
    }
}
